File: scripts/research_run.py
#!/usr/bin/env python3
"""Fail-closed experiment launcher with a hash-bound contract and ledger.

The manifest is accepted and its command launched in one process. There is no
separate "prepare" state in which unrecorded diagnostics can be run. Every
finished command receives one verdict and a hash-chained ledger entry.
"""

from __future__ import annotations

import hashlib
import json
import os
import subprocess
import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
CONTRACT = ROOT / "research" / "contract.json"
LEDGER = ROOT / "research" / "ledger.jsonl"
ACTIVE = ROOT / "research" / "active"
REQUIRED = {
    "experiment_id", "contract_id", "contract_sha256", "stage",
    "bottleneck", "hypothesis", "prediction", "likely_failure",
    "single_change", "baseline", "evaluation_population",
    "rejection_rule", "promotion_rule", "evidence_class", "command",
    "output",
}
ALLOWED_VERDICTS = {
    "diagnostic_only", "smoke_only", "inconclusive", "rejected",
    "confirmation_pending", "accepted", "blocked",
}


def sha256(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def load_json(path: Path) -> dict:
    value = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(value, dict):
        raise ValueError(f"{path} must contain one JSON object")
    return value


def prior_entries() -> list[dict]:
    if not LEDGER.exists():
        return []
    entries = []
    previous = "0" * 64
    for line_number, line in enumerate(LEDGER.read_text(encoding="utf-8").splitlines(), 1):
        entry = json.loads(line)
        digest = entry.pop("entry_sha256", None)
        if entry.get("previous_entry_sha256") != previous:
            raise ValueError(f"ledger hash chain breaks at line {line_number}")
        expected = hashlib.sha256(json.dumps(
            entry, sort_keys=True, separators=(",", ":")).encode()).hexdigest()
        if digest != expected:
            raise ValueError(f"ledger entry hash differs at line {line_number}")
        entry["entry_sha256"] = digest
        entries.append(entry)
        previous = digest
    return entries


def append_receipt(receipt: dict, prior: list[dict]) -> None:
    receipt["previous_entry_sha256"] = (
        prior[-1]["entry_sha256"] if prior else "0" * 64)
    encoded = json.dumps(receipt, sort_keys=True, separators=(",", ":"))
    receipt["entry_sha256"] = hashlib.sha256(encoded.encode()).hexdigest()
    LEDGER.parent.mkdir(parents=True, exist_ok=True)
    line = json.dumps(receipt, sort_keys=True, separators=(",", ":")) + "\n"
    descriptor = os.open(LEDGER, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o644)
    try:
        os.write(descriptor, line.encode())
        os.fsync(descriptor)
    finally:
        os.close(descriptor)


def validate(manifest_path: Path) -> tuple[dict, dict, list[dict]]:
    contract = load_json(CONTRACT)
    manifest = load_json(manifest_path)
    missing = REQUIRED - manifest.keys()
    if missing:
        raise ValueError(f"manifest is missing: {', '.join(sorted(missing))}")
    if manifest["contract_id"] != contract["contract_id"]:
        raise ValueError("manifest names another research contract")
    if manifest["contract_sha256"] != sha256(CONTRACT):
        raise ValueError("manifest is stale: contract SHA-256 differs")
    if manifest["stage"] != contract["current_stage"]:
        raise ValueError("manifest does not address the earliest unresolved stage")
    if manifest["evidence_class"] not in ALLOWED_VERDICTS - {"accepted", "blocked"}:
        raise ValueError("manifest asks for an invalid predeclared evidence class")
    command = manifest["command"]
    if (not isinstance(command, list) or not command
            or not all(isinstance(item, str) and item for item in command)):
        raise ValueError("manifest command must be a non-empty string array")
    output = (ROOT / manifest["output"]).resolve()
    if ROOT not in output.parents:
        raise ValueError("experiment output must stay inside the QOMM workspace")
    prior = prior_entries()
    if any(row["experiment_id"] == manifest["experiment_id"] for row in prior):
        raise ValueError("experiment identifier already has a terminal receipt")
    return contract, manifest, prior


def run(manifest_path: Path) -> int:
    _, manifest, prior = validate(manifest_path)
    ACTIVE.mkdir(parents=True, exist_ok=True)
    active = ACTIVE / f"{manifest['experiment_id']}.json"
    descriptor = os.open(active, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    try:
        os.write(descriptor, manifest_path.read_bytes())
        os.fsync(descriptor)
    finally:
        os.close(descriptor)

    output = (ROOT / manifest["output"]).resolve()
    output.parent.mkdir(parents=True, exist_ok=True)
    started = time.time_ns()
    completed = None
    error = None
    try:
        completed = subprocess.run(manifest["command"], cwd=ROOT, check=False,
                                   capture_output=True, text=True)
        if completed.returncode != 0:
            error = (completed.stderr or completed.stdout)[-4000:]
        elif not output.is_file():
            error = "declared output was not created"
    except Exception as exc:
        error = f"{type(exc).__name__}: {exc}"

    verdict = manifest["evidence_class"] if error is None else "blocked"
    receipt = {
        "experiment_id": manifest["experiment_id"],
        "contract_id": manifest["contract_id"],
        "contract_sha256": manifest["contract_sha256"],
        "manifest": str(manifest_path.resolve().relative_to(ROOT)),
        "manifest_sha256": sha256(manifest_path),
        "started_at_ns": started,
        "elapsed_seconds": (time.time_ns() - started) / 1_000_000_000,
        "verdict": verdict,
        "return_code": None if completed is None else completed.returncode,
        "output": manifest["output"],
        "output_sha256": sha256(output) if error is None else None,
        "error": error,
    }
    append_receipt(receipt, prior)
    active.unlink(missing_ok=True)
    if error is not None:
        print(error, file=sys.stderr)
        return 1
    print(json.dumps(receipt, sort_keys=True))
    return 0


def main() -> int:
    if len(sys.argv) != 2:
        raise SystemExit("usage: research_run.py MANIFEST.json")
    return run(Path(sys.argv[1]).resolve())


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/qomm-harness/src/bin/research_run.rs
//! Fail-closed experiment launcher with a hash-bound contract and ledger.

use qomm_harness::rust_only::{validate_experiment_command, validate_repository};
use qomm_harness::{repo_root, HarnessResult};
use serde_json::{Map, Value};
use sha2::{Digest, Sha256};
use std::collections::BTreeSet;
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::os::unix::fs::OpenOptionsExt;
use std::path::{Component, Path, PathBuf};
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

type ValidatedResearch = (Map<String, Value>, Vec<Map<String, Value>>);

const REQUIRED: [&str; 16] = [
    "experiment_id",
    "contract_id",
    "contract_sha256",
    "stage",
    "bottleneck",
    "hypothesis",
    "prediction",
    "likely_failure",
    "single_change",
    "baseline",
    "evaluation_population",
    "rejection_rule",
    "promotion_rule",
    "evidence_class",
    "command",
    "output",
];

const ALLOWED_VERDICTS: [&str; 7] = [
    "diagnostic_only",
    "smoke_only",
    "inconclusive",
    "rejected",
    "confirmation_pending",
    "accepted",
    "blocked",
];

struct Paths {
    root: PathBuf,
    contract: PathBuf,
    ledger: PathBuf,
    active: PathBuf,
}

impl Paths {
    fn discover() -> HarnessResult<Self> {
        let root = std::env::var_os("QOMM_RESEARCH_ROOT")
            .map(PathBuf::from)
            .unwrap_or_else(repo_root)
            .canonicalize()?;
        Ok(Self {
            contract: root.join("research/contract.json"),
            ledger: root.join("research/ledger.jsonl"),
            active: root.join("research/active"),
            root,
        })
    }
}

fn sha256(path: &Path) -> HarnessResult<String> {
    Ok(hex::encode(Sha256::digest(fs::read(path)?)))
}

fn load_json(path: &Path) -> HarnessResult<Map<String, Value>> {
    let value: Value = serde_json::from_slice(&fs::read(path)?)?;
    value
        .as_object()
        .cloned()
        .ok_or_else(|| format!("{} must contain one JSON object", path.display()).into())
}

fn canonical_json(value: &Value) -> String {
    render_json(value, false)
}

fn spaced_json(value: &Value) -> String {
    render_json(value, true)
}

fn render_json(value: &Value, spaces: bool) -> String {
    match value {
        Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_) => {
            serde_json::to_string(value).expect("JSON scalar serializes")
        }
        Value::Array(values) => {
            let separator = if spaces { ", " } else { "," };
            format!(
                "[{}]",
                values
                    .iter()
                    .map(|value| render_json(value, spaces))
                    .collect::<Vec<_>>()
                    .join(separator)
            )
        }
        Value::Object(values) => {
            let item_separator = if spaces { ", " } else { "," };
            let key_separator = if spaces { ": " } else { ":" };
            format!(
                "{{{}}}",
                values
                    .iter()
                    .map(|(key, value)| format!(
                        "{}{key_separator}{}",
                        serde_json::to_string(key).expect("JSON key serializes"),
                        render_json(value, spaces)
                    ))
                    .collect::<Vec<_>>()
                    .join(item_separator)
            )
        }
    }
}

fn prior_entries(paths: &Paths) -> HarnessResult<Vec<Map<String, Value>>> {
    if !paths.ledger.exists() {
        return Ok(Vec::new());
    }
    let mut entries = Vec::new();
    let mut previous = "0".repeat(64);
    for (line_number, line) in fs::read_to_string(&paths.ledger)?.lines().enumerate() {
        let mut entry = serde_json::from_str::<Value>(line)?
            .as_object()
            .cloned()
            .ok_or_else(|| format!("ledger line {} is not an object", line_number + 1))?;
        let digest = entry
            .remove("entry_sha256")
            .and_then(|value| value.as_str().map(str::to_owned));
        if entry.get("previous_entry_sha256").and_then(Value::as_str) != Some(previous.as_str()) {
            return Err(format!("ledger hash chain breaks at line {}", line_number + 1).into());
        }
        let expected = hex::encode(Sha256::digest(canonical_json(&Value::Object(
            entry.clone(),
        ))));
        if digest.as_deref() != Some(expected.as_str()) {
            return Err(format!("ledger entry hash differs at line {}", line_number + 1).into());
        }
        entry.insert("entry_sha256".into(), Value::String(expected.clone()));
        entries.push(entry);
        previous = expected;
    }
    Ok(entries)
}

fn append_receipt(
    paths: &Paths,
    receipt: &mut Map<String, Value>,
    prior: &[Map<String, Value>],
) -> HarnessResult<()> {
    let previous = prior
        .last()
        .and_then(|entry| entry.get("entry_sha256"))
        .and_then(Value::as_str)
        .map(str::to_owned)
        .unwrap_or_else(|| "0".repeat(64));
    receipt.insert("previous_entry_sha256".into(), Value::String(previous));
    let encoded = canonical_json(&Value::Object(receipt.clone()));
    let digest = hex::encode(Sha256::digest(encoded));
    receipt.insert("entry_sha256".into(), Value::String(digest));
    if let Some(parent) = paths.ledger.parent() {
        fs::create_dir_all(parent)?;
    }
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .mode(0o644)
        .open(&paths.ledger)?;
    writeln!(file, "{}", canonical_json(&Value::Object(receipt.clone())))?;
    file.sync_all()?;
    Ok(())
}

fn resolve_manifest(argument: &Path) -> HarnessResult<PathBuf> {
    if argument.is_absolute() {
        Ok(argument.canonicalize()?)
    } else {
        Ok(std::env::current_dir()?.join(argument).canonicalize()?)
    }
}

fn resolve_output(root: &Path, raw: &str) -> HarnessResult<PathBuf> {
    let relative = Path::new(raw);
    if relative.is_absolute() {
        return Err("experiment output must stay inside the QOMM workspace".into());
    }
    let mut output = root.to_path_buf();
    for component in relative.components() {
        match component {
            Component::Normal(value) => output.push(value),
            Component::CurDir => {}
            Component::ParentDir => {
                if output == root || !output.pop() {
                    return Err("experiment output must stay inside the QOMM workspace".into());
                }
            }
            Component::RootDir | Component::Prefix(_) => {
                return Err("experiment output must stay inside the QOMM workspace".into());
            }
        }
    }
    if !output.starts_with(root) || output == root {
        return Err("experiment output must stay inside the QOMM workspace".into());
    }
    Ok(output)
}

fn validate(paths: &Paths, manifest_path: &Path) -> HarnessResult<ValidatedResearch> {
    validate_repository(&paths.root)?;
    let contract = load_json(&paths.contract)?;
    let manifest = load_json(manifest_path)?;
    let missing = REQUIRED
        .iter()
        .copied()
        .collect::<BTreeSet<_>>()
        .difference(&manifest.keys().map(String::as_str).collect::<BTreeSet<_>>())
        .copied()
        .collect::<Vec<_>>();
    if !missing.is_empty() {
        return Err(format!("manifest is missing: {}", missing.join(", ")).into());
    }
    if manifest.get("contract_id") != contract.get("contract_id") {
        return Err("manifest names another research contract".into());
    }
    let contract_sha256 = sha256(&paths.contract)?;
    if manifest.get("contract_sha256").and_then(Value::as_str) != Some(contract_sha256.as_str()) {
        return Err("manifest is stale: contract SHA-256 differs".into());
    }
    if manifest.get("stage") != contract.get("current_stage") {
        return Err("manifest does not address the earliest unresolved stage".into());
    }
    let evidence = manifest
        .get("evidence_class")
        .and_then(Value::as_str)
        .unwrap_or("");
    if !ALLOWED_VERDICTS[..5].contains(&evidence) {
        return Err("manifest asks for an invalid predeclared evidence class".into());
    }
    let command = manifest.get("command").and_then(Value::as_array);
    if command.is_none_or(|command| {
        command.is_empty()
            || command
                .iter()
                .any(|item| item.as_str().is_none_or(str::is_empty))
    }) {
        return Err("manifest command must be a non-empty string array".into());
    }
    let command = command.expect("command was just validated");
    validate_experiment_command(
        &command
            .iter()
            .map(|item| item.as_str().expect("command item was just validated"))
            .collect::<Vec<_>>(),
    )?;
    let raw_output = manifest
        .get("output")
        .and_then(Value::as_str)
        .ok_or("manifest output must be a string")?;
    resolve_output(&paths.root, raw_output)?;
    let prior = prior_entries(paths)?;
    let experiment_id = manifest
        .get("experiment_id")
        .and_then(Value::as_str)
        .unwrap_or("");
    if prior
        .iter()
        .any(|row| row.get("experiment_id").and_then(Value::as_str) == Some(experiment_id))
    {
        return Err("experiment identifier already has a terminal receipt".into());
    }
    Ok((manifest, prior))
}

fn unix_nanos() -> u128 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos()
}

fn run(paths: &Paths, manifest_path: &Path) -> HarnessResult<i32> {
    let (manifest, prior) = validate(paths, manifest_path)?;
    fs::create_dir_all(&paths.active)?;
    let experiment_id = manifest["experiment_id"]
        .as_str()
        .ok_or("experiment_id must be a string")?;
    let active = paths.active.join(format!("{experiment_id}.json"));
    let mut active_file = OpenOptions::new()
        .write(true)
        .create_new(true)
        .mode(0o600)
        .open(&active)?;
    active_file.write_all(&fs::read(manifest_path)?)?;
    active_file.sync_all()?;
    drop(active_file);

    let raw_output = manifest["output"]
        .as_str()
        .ok_or("manifest output must be a string")?;
    let output = resolve_output(&paths.root, raw_output)?;
    if let Some(parent) = output.parent() {
        fs::create_dir_all(parent)?;
    }
    let started = unix_nanos();
    let command = manifest["command"]
        .as_array()
        .expect("command was validated")
        .iter()
        .map(|value| value.as_str().expect("command item was validated"))
        .collect::<Vec<_>>();
    let mut return_code = None;
    let mut error = None::<String>;
    match Command::new(command[0])
        .args(&command[1..])
        .current_dir(&paths.root)
        .output()
    {
        Ok(completed) => {
            return_code = completed.status.code();
            if !completed.status.success() {
                let stderr = String::from_utf8_lossy(&completed.stderr);
                let stdout = String::from_utf8_lossy(&completed.stdout);
                error = Some(tail(
                    if stderr.is_empty() { &stdout } else { &stderr },
                    4_000,
                ));
            } else if !output.is_file() {
                error = Some("declared output was not created".into());
            }
        }
        Err(failure) => error = Some(format!("{}: {failure}", io_error_name(&failure))),
    }
    let finished = unix_nanos();
    let verdict = if error.is_none() {
        manifest["evidence_class"].clone()
    } else {
        Value::String("blocked".into())
    };
    let mut receipt = Map::from_iter([
        ("experiment_id".into(), manifest["experiment_id"].clone()),
        ("contract_id".into(), manifest["contract_id"].clone()),
        (
            "contract_sha256".into(),
            manifest["contract_sha256"].clone(),
        ),
        (
            "manifest".into(),
            Value::String(
                manifest_path
                    .strip_prefix(&paths.root)?
                    .to_string_lossy()
                    .into_owned(),
            ),
        ),
        (
            "manifest_sha256".into(),
            Value::String(sha256(manifest_path)?),
        ),
        ("started_at_ns".into(), Value::from(started as u64)),
        (
            "elapsed_seconds".into(),
            Value::from((finished - started) as f64 / 1_000_000_000.0),
        ),
        ("verdict".into(), verdict),
        (
            "return_code".into(),
            return_code.map_or(Value::Null, Value::from),
        ),
        ("output".into(), manifest["output"].clone()),
        (
            "output_sha256".into(),
            if error.is_none() {
                Value::String(sha256(&output)?)
            } else {
                Value::Null
            },
        ),
        (
            "error".into(),
            error.clone().map_or(Value::Null, Value::String),
        ),
    ]);
    append_receipt(paths, &mut receipt, &prior)?;
    match fs::remove_file(&active) {
        Ok(()) => {}
        Err(failure) if failure.kind() == std::io::ErrorKind::NotFound => {}
        Err(failure) => return Err(failure.into()),
    }
    if let Some(error) = error {
        eprintln!("{error}");
        return Ok(1);
    }
    println!("{}", spaced_json(&Value::Object(receipt)));
    Ok(0)
}

fn io_error_name(error: &std::io::Error) -> &'static str {
    match error.kind() {
        std::io::ErrorKind::NotFound => "FileNotFoundError",
        std::io::ErrorKind::PermissionDenied => "PermissionError",
        _ => "OSError",
    }
}

fn tail(text: &str, chars: usize) -> String {
    let values = text.chars().collect::<Vec<_>>();
    values[values.len().saturating_sub(chars)..]
        .iter()
        .collect()
}

fn main() {
    let args = std::env::args_os().skip(1).collect::<Vec<_>>();
    if args.len() != 1 {
        eprintln!("usage: research_run MANIFEST.json");
        std::process::exit(1);
    }
    let result = (|| -> HarnessResult<i32> {
        let paths = Paths::discover()?;
        let manifest = resolve_manifest(Path::new(&args[0]))?;
        run(&paths, &manifest)
    })();
    match result {
        Ok(code) => std::process::exit(code),
        Err(error) => {
            eprintln!("research_run: {error}");
            std::process::exit(1);
        }
    }
}
